### ISSUE TITLE
Directive binding workflow has changed in version 1.4.1

### ISSUE THREAD

**Issue by twolff-iow** [angular/angular.js#12192]
Hi,

I have recently updated to version 1.4.1 and some of my custom directives are now showing a different behaviour and I would like to know whether this is expected and ask for an advise.

In one of my custom directives, I define optional attributes like this:
```
        return {
            restrict: 'E',
            ...
            scope: {
                state: '@?',
                type: '@?'
            },
           ...
       }
```
and have set default values so far as this:
```
            link: function (scope, element, attrs) {
                if (attrs.state === undefined) {
                    scope.state = false;
                }
                if (attrs.type === undefined) {
                    scope.type = 'default';
                }
            }
```
This is properly working with version 1.4.0, but fails to work in version 1.4.1 now. Comparing the two versions I found the following addition in 1.4.1 in "initializeDirectiveBindings" method:
```
        if (!hasOwnProperty.call(attrs, attrName)) {
          // In the case of user defined a binding with the same name as a method in Object.prototype but didn't set
          // the corresponding attribute. We need to make sure subsequent code won't access to the prototype function
          attrs[attrName] = undefined;
        }
```
The initialization of the array value makes the "$observe" function fire the following asynchronous evaluation:
```
        $rootScope.$evalAsync(function() {
          if (!listeners.$$inter && attrs.hasOwnProperty(key)) {
            // no one registered attribute interpolation function, so lets call it manually
            fn(attrs[key]);
          }
        });
```
as the condition "attrs.hasOwnProperty(key)" is now true which was not the case in version 1.4.0 due to the absence of the initialization above. The call "fn(attrs[key])" sets the scope variable to the attribute value.

Due to the asynchronicity these calls are made after me setting the default values and hence, the scope variables become undefined. If I move the setting of the default values with a $timeout into the next cycle, it works again:
```
            link: function (scope, element, attrs) {
                $timeout(function () {
                    if (attrs.state === undefined) {
                        scope.state = false;
                    }
                    if (attrs.type === undefined) {
                        scope.type = 'default';
                    }
                });
            }
```
Alternatively, I can tweak the "attrs.*" values liek below, as this way the $observer will take care to push the new values into the scope:
```
            link: function (scope, element, attrs) {
                if (attrs.state === undefined) {
                    attrs.state = false;
                }
                if (attrs.type === undefined) {
                    attrs.type = 'default';
                }
            }
```
Can you let me know whether this was expected? Please also let me know if the way setting default values should be done differently.

Here is a plunker for version 1.4.0: http://plnkr.co/edit/32BXMfj7SCnyRSj8aoeD?p=preview
And a fork for version 1.4.1: http://plnkr.co/edit/V8wioTB281oGB7o3FmDj?p=preview

Thanks,
  Tobias.

**Comment by gkalpak** [comment]
I don't see any difference between the two plnkrs.
What is supposed to be happening differently with 1.4.1 ?

<sub>BTW, there was a similar issue with directive controllers (which has been fixed with 8a1eb160), but it shouldn't affect your usecase.</sub>

**Comment by twolff-iow** [comment]
Sorry, for not making this clear. In the version 1.4.0, the Plunkr shows "default" for the first directive usage and "test" for the second. When you look into the 1.4.1 Plunkr, you do see the "test" in the second directive, but you do NOT see "default" in the first one.

**Comment by gkalpak** [comment]
Well, I see "default" in both plnkrs :confused:

**Comment by twolff-iow** [comment]
No kidding! I now see the same, but I swear, it has been different 2 hours ago. I would not have taken the effort setting this up with no reason.

**Comment by twolff-iow** [comment]
Just found the issue. I had been trying workarounds on the Plunkrs already. Give me a minute and I will provide new ones. Sorry for the hickups.

**Comment by twolff-iow** [comment]
Here is another fork: http://plnkr.co/edit/RYUBSU864W8U3E9C6oLb

I accidently had "attrs.value = 'default'" in the code which should have read "scope.value = 'default'".

**Comment by gkalpak** [comment]
This has been fixed (8a1eb16) and [works on master][1].
The fix will be available in 1.4.2.

--
In the mean time, you can change `scope.value = 'default'` to `attrs.value = 'default'` (or have both if you need scope to be updated synchronously).

[1]: http://plnkr.co/edit/iY1hWbJZ9ImUK5iebfUz?p=preview
